pandas/tests/indexes/interval/test_astype.py

```diff
@@ -153,22 +155,15 @@ def test_subtype_integer(self, subtype):
         with pytest.raises(ValueError, match=msg):
             index.insert(0, np.nan).astype(dtype)
 
-    @pytest.mark.xfail(reason="GH#15832")
     def test_subtype_integer_errors(self):
         # float64 -> uint64 fails with negative values
         index = interval_range(-10.0, 10.0)
         dtype = IntervalDtype("uint64")
-        with pytest.raises(ValueError):
-            index.astype(dtype)
-
-        # float64 -> integer-like fails with non-integer valued floats
-        index = interval_range(0.0, 10.0, freq=0.25)
-        dtype = IntervalDtype("int64")
-        with pytest.raises(ValueError):
-            index.astype(dtype)
-
-        dtype = IntervalDtype("uint64")
-        with pytest.raises(ValueError):
+        msg = re.escape(
+            "Cannot convert interval[float64] to interval[uint64]; subtypes are "
+            "incompatible"
+        )
+        with pytest.raises(TypeError, match=msg):
             index.astype(dtype)
```

Comment:
This is a test that was marked as failing, with the reason being closed issue #15832 . I put a bit of discussion about this over in the issue.

I removed the `xfail` and I also removed the last two `pytest.raises`. The result of the coercion is: `IntervalIndex([(0, 0], (0, 0], (0, 0], (0, 1], (1, 1] ... (8, 9], (9, 9], (9, 9], (9, 9], (9, 10]],  closed='right',  dtype='interval[int64]')` which doesn't look particularly useful but isn't exactly wrong either. If I understand the discussion at #15832 correctly this is intended behaviour for it not to remove the error.

Reply:
Looks like the test for 
```
index = interval_range(0.0, 10.0, freq=0.25)
dtype = IntervalDtype("int64")
```
is lost here?

Reply:
Yes, I removed it deliberately, as well as the equivalent one for `uint64`. I tried to explain that in my comment above but I guess I wasn't clear.

This test was marked as failing before. Now it is not. The test, as written before, was saying that 
```
index = interval_range(0.0, 10.0, freq=0.25)
dtype = IntervalDtype("int64")
```
should raise a `ValueError`. It does not raise any error. The question here is: Should casting an interval range with non-integer float boundaries raise an error? This test said it should, but it was marked as failing, because right now it does not raise an error. If I understood the discussion at #15832 correctly, and it is possible that I did not, than I think that it should not raise an error, so this test is not required.

The question is, as a user of pandas, if you had an interval range such as `[(0.0, 0.25], (0.25, 0.5], (0.5, 0.75], (0.75, 1.0], (1.0, 1.25], (1.25, 1.5], (1.5, 1.75], (1.75, 2.0], (2.0, 2.25], (2.25, 2.5]]`, and you cast it to int, what would you want or expect? The old (failing) version of the test said you expect a ValueError. I am saying you expect `[(0, 0], (0, 0], (0, 0], (0, 1], (1, 1], (1, 1], (1, 1], (1, 2], (2, 2], (2, 2]]` which, as I said, looks not that useful as an IntervalIndex, but is not wrong either.

Would love to hear your thoughts.

I could also add another test for the expected behaviour of casting from float to int intervals if that's what we want.

Reply:
Thanks for explaining, sorry should've read your comment on that linked issue. Your logic makes sense to me, but I'd defer to someone with more experience on how we handle `IntervalIndex`. 

If being removed from this test, we definitely want to make sure this specific casting behavior is being tested somewhere.  To keep this pr simple and easier to merge, I'd suggest reverting modification of this test, then making a followup pr to clean up this test and add tests for any untested casting behavior.

Reply:
Ok, but I am not quite sure how to handle it. I want to leave the code in a good state even if I plan to do another PR soon, because it's possible that one will take a while or not be accepted ever or whatever. The way I see it, these are our options:

1. Revert it to exactly what it was before I modified it. That would make the file fail the new linting rule we want to add as part of #30999 , so don't add the linting rule and leave that issue open. I'm not particularly happy with this option because I have gotten the number of bare pytest.raises down from 250ish to 64 in just a few days and I was hoping to finish in a few more days. And the longer #30999 stays open the more likely people are to add new bare pytest.raises and the issue will stay open forever.
2. Revert it, and simultaneously add to the linting script a check that the bare pytest.raises is not in test decorated with pytest xfail. Generally, addressing #30999 has been a pretty mechanical process, except when it comes to tests marked as failures. Then it takes a lot of digging through the code and historical issues to understand what would be a the correct failure message to add. It's the one thing that maybe makes #30999 not as good as a "good first issue" as it appears. I expect to run into more such issues as I try to finish all the modules, because there are 26 files that only have 1 or 2 bare pytest raises and I suspect it's because they are the hard ones that people didn't know what message to put in. Adding the check to the script seems like kinda a cool project to figure out the ast module better than I currently do, but probably just that and not really worth our time. Unless we really think that we want the linter to do that.
3. Come up with, out of whole cloth, the error message that I think should be included in such a `ValueError`, if in the future pandas were to raise one. This is the strategy I took in previous pull requests. (see #38609 for example, where I added a new completely invented error message to `test_multiple_open_close`). This is also what I was trying to do that led me to read the code and the past issues and decide that I didn't think it deserved an error message at all and therefore that the test didn't make sense and I would remove it.
4. Put in an obviously completely bogus error message, so that future developers know that it's not a real one, and whoever in the future decides to address that test either by making it pass or by removing it can write a real message in the future.

Anyway, after thinking it through and writing it out, I think I will try for strategy 3. but let me know if there's a different one that you prefer.

Apologies for the wall of text; I am new to contributing to pandas and obviously spent a lot of time thinking about this.

Reply:
ok done

Reply:
xref https://github.com/pandas-dev/pandas/issues/38718

Reply:
New PR #38719

Reply:
If that's been done in past PR's, probably fine here too. Appreciate the thoughtfulness on your approach here!

Reply:
yeah generally addressing on a new PR is the correct way, will look at that soon.